### PR TITLE
feat(MULTI): Introduce support for Apollo

### PR DESCRIPTION
**Pull request by pierlo-upitup**
This PR introduces support for using [`apollo-client`](https://github.com/apollographql/apollo-client)  (1.x as 2.x is unstable) .

TODO
- [x] add support for ApolloProvider without requiring a Redux `store`
- [x] add an example to the complex app @pierlo-upitup
- [x] add documentation about how to implement an `apollo.js` module @pierlo-upitup
- [x] ~~depends on upcoming React 16 upgrade (cc @dlmr)~~

### CODE REVIEW

File: packages/roc-package-web-app-react/package.json [hunk 1]
```diff
@@ -25,13 +25,16 @@
...
+    "apollo-client": "^1.9.3",
     "debug": "~2.2.0",
     "error": "~7.0.2",
+    "graphql": "^0.10.0",
     "history": "^3.0.0",
     "intl": "^1.2.5",
     "intl-locales-supported": "^1.0.0",
     "nunjucks": "~2.4.2",
     "pretty-error": "^2.1.0",
+    "react-apollo": "^1.4.15",
```
Comment (dlmr) [inline packages/roc-package-web-app-react/package.json]: This should point to `^1.4.16`

File: packages/roc-package-web-app-react/app/client/create-client.js [hunk 2]
```diff
@@ -96,12 +97,46 @@ export default function createClient({
         };
         const createComponent = [(component) => component];
         const createDevComponent = [(component) => component];
+        if (HAS_APOLLO && !HAS_REDUX_REDUCERS) {
+            const { ApolloProvider } = require('react-apollo');
+            const { default: ApolloClient, createNetworkInterface } = require('apollo-client');
```
Comment (dlmr) [inline packages/roc-package-web-app-react/app/client/create-client.js]: Let's remove this import.

File: packages/roc-package-web-app-react/app/client/create-client.js [hunk 3]
```diff
@@ -96,12 +97,46 @@ export default function createClient({
         };
         const createComponent = [(component) => component];
         const createDevComponent = [(component) => component];
+        if (HAS_APOLLO && !HAS_REDUX_REDUCERS) {
+            const { ApolloProvider } = require('react-apollo');
```
Comment (dlmr) [inline packages/roc-package-web-app-react/app/client/create-client.js]: Let's change this import to be
```js
const { ApolloProvider, ApolloClient, createNetworkInterface } = require('react-apollo');
```

File: packages/roc-package-web-app-react/app/client/create-client.js [hunk 4]
```diff
@@ -96,12 +97,46 @@ export default function createClient({
...
+            ));
+        } else if (HAS_REDUX_REDUCERS && createStore) {
+            let Provider;
+            const extraMiddlewares = [];
+            const extraReducers = {};
+            const providerProps = {};
+
+            if (HAS_APOLLO) {
+                Provider = require('react-apollo').ApolloProvider;
+                const { default: ApolloClient, createNetworkInterface } = require('apollo-client');
```
Comment (dlmr) [inline packages/roc-package-web-app-react/app/client/create-client.js]: Let's change this
```js
const { ApolloProvider, ApolloClient, createNetworkInterface } = require('react-apollo');
Provider = ApolloProvider;
```

File: packages/roc-package-web-app-react/package.json [hunk 5]
```diff
@@ -25,13 +25,16 @@
   ],
   "license": "MIT",
   "dependencies": {
+    "apollo-client": "^1.9.3",
```
Comment (dlmr) [inline packages/roc-package-web-app-react/package.json]: We should remove this dependency.

File: packages/roc-package-web-app-react/src/app/server/setupForRender.js [hunk 6]
```diff
@@ -2,22 +2,47 @@
 
 import { createMemoryHistory } from 'react-router';
 import { syncHistoryWithStore } from 'react-router-redux';
+import { ApolloClient, createNetworkInterface } from 'apollo-client';
```
Comment (dlmr) [inline packages/roc-package-web-app-react/src/app/server/setupForRender.js]: We should import this from `react-apollo`.

File: packages/roc-package-web-app-react/src/app/server/reactRenderer.js [hunk 7]
```diff
@@ -208,48 +210,85 @@ export function reactRender({
...
-                    if (store) {
+                    if (apollo && !store) {
+                        component = <ApolloProvider client={apollo}>{component}</ApolloProvider>;
+                    } else if (apollo && store) {
+                        component = (
+                            <ApolloProvider store={store} client={apollo}>
+                                {component}
+                            </ApolloProvider>
+                        );
+                    } else if (store) {
```
Comment (dlmr) [inline packages/roc-package-web-app-react/src/app/server/reactRenderer.js]: I think we can change this to
```jsx
if (apollo) {
    const providerProps = { client: apollo, store };
    component = <ApolloProvider {...providerProps}>{component}</ApolloProvider>;
} else if (store) {
    component = <Provider store={store}>{component}</Provider>;
}
```

File: packages/roc-package-web-app-react/src/roc/index.js [hunk 8]
```diff
@@ -54,9 +54,12 @@ export default {
     dependencies: {
         exports: {
             ...generateDependencies(packageJSON, [
+                'apollo-client',
```
Comment (dlmr) [inline packages/roc-package-web-app-react/src/roc/index.js]: We can remove this.

**Comment by dlmr** [comment]
Seems like we don't have a hard dependency on React 16 here, and that both 15 and 16 will work.

**Comment by pierlo-upitup** [comment]
Force-pushed the amended commit with feedback addressed.

**Comment by pierlo-upitup** [comment]
Complex example app now adds tabs to switch between GraphQL and Redux/Fetch. 
The GraphQL endpoint uses https://www.graphqlhub.com/playground to wrap GitHub's RESTful API (the GH GraphQL endpoints require auth, which would make things unnecessarily complex).

 image

**Pull request merged**